src/react/hooks/use-presence.ts

```diff
@@ -0,0 +1,160 @@
+  /**
+   * A shortcut to the {@link Presence.update} method.
+   */
+  readonly update: Presence['update'];
+
+  /**
+   * A shortcut to the {@link Presence.leave} method.
+   */
+  readonly leave: Presence['leave'];
+
+  /**
+   * Indicates whether the current user is present in the room.
+   */
+  readonly isPresent: boolean;
+
+  /**
+   * Indicates if an error occurred while entering or leaving the room.
+   */
+  readonly error?: Ably.ErrorInfo;
+
+  /**
+   * Provides access to the underlying {@link Presence} instance of the room.
+   */
+  readonly presence: Presence;
+}
+
+const INACTIVE_CONNECTION_STATES = new Set<ConnectionLifecycle>([
+  ConnectionLifecycle.Suspended,
+  ConnectionLifecycle.Failed,
+]);
+
+/**
+ * A hook that provides access to the {@link Presence} instance in the room.
+ * It will use the instance belonging to the room in the nearest {@link RoomProvider} in the component tree.
+ * On calling, the hook will enter the room with the provided data and leave the room when the component unmounts.
+ * The {@link isPresent} flag will indicate when the user has become present in the room.
+ *
+ * @param params - Allows the registering of optional callbacks.
+ * @returns UsePresenceResponse - An object containing the {@link Presence} instance and methods to interact with it.
+ */
+export const usePresence = (params?: UsePresenceParams): UsePresenceResponse => {
+  const { currentStatus: connectionStatus, error: connectionError } = useChatConnection({
+    onStatusChange: params?.onConnectionStatusChange,
+  });
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const [isPresent, setIsPresent] = useState(false);
+  const [error, setError] = useState<Ably.ErrorInfo | undefined>();
+
+  // store the roomStatus in a ref to ensure the correct value is used in the effect cleanup
+  const roomStatusAndConnectionStatusRef = useRef({ roomStatus, connectionStatus });
+
+  // we can't use the data param directly in a dependency array as it will cause an infinite loop
+  const dataRef = useRef(params?.data);
+  useEffect(() => {
+    dataRef.current = params?.data;
+  }, [params?.data]);
+
+  useEffect(() => {
+    // Update the ref when roomStatus changes
+    roomStatusAndConnectionStatusRef.current = { roomStatus, connectionStatus };
+  }, [roomStatus, connectionStatus]);
+
+  // enter the room when the hook is mounted
+  useEffect(() => {
+    const canJoinPresence = roomStatus === RoomLifecycle.Attached && !INACTIVE_CONNECTION_STATES.has(connectionStatus);
+    const canLeavePresence =
+      roomStatusAndConnectionStatusRef.current.roomStatus === RoomLifecycle.Attached &&
+      !INACTIVE_CONNECTION_STATES.has(roomStatusAndConnectionStatusRef.current.connectionStatus);
+
+    // wait until the room is attached before attempting to enter, and ensure the connection is active
+    if (!canJoinPresence) return;
+    room.presence
```

Comment:
I've seen some discussion before -- ignore if already resolved. We enter presence by default when this element is in scope? We can't just use it to list the users who are in presence without joining?

I think there's some room for use cases where you're not allowed to join presence (not VIP?) but you can see who's there (the list of VIPs)?

Reply:
That's correct, @AndyTWF and I discussed this and he thought the expectation here by users would be the hook automatically joins presence.